File: mcp-code-intelligence-nodejs/src/ollama/ollama-client.ts
/**
 * Ollama HTTP Client — Communicates with local Ollama API for embeddings.
 * Gracefully degrades if Ollama is not available.
 */

export interface EmbeddingResponse {
  embedding: number[];
}

export interface OllamaConfig {
  url: string;
  model: string;
}

export class OllamaClient {
  private url: string;
  private model: string;
  private available: boolean | null = null;

  constructor(config: OllamaConfig) {
    this.url = config.url;
    this.model = config.model;
  }

  /** Check if Ollama is reachable. */
  async isAvailable(): Promise<boolean> {
    if (this.available !== null) return this.available;
    try {
      const resp = await fetch(`${this.url}/api/tags`, { signal: AbortSignal.timeout(3000) });
      this.available = resp.ok;
    } catch {
      this.available = false;
    }
    console.error(`[ollama] Available: ${this.available}`);
    return this.available;
  }

  /** Generate embedding vector for text. */
  async generateEmbedding(text: string): Promise<number[] | null> {
    if (!await this.isAvailable()) return null;
    try {
      const resp = await fetch(`${this.url}/api/embeddings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model: this.model, prompt: text }),
        signal: AbortSignal.timeout(30000),
      });
      if (!resp.ok) return null;
      const data = await resp.json() as EmbeddingResponse;
      return data.embedding;
    } catch (err) {
      console.error('[ollama] Embedding error:', err);
      return null;
    }
  }

  /** Generate embeddings for multiple texts in batch. */
  async generateBatch(texts: string[]): Promise<(number[] | null)[]> {
    const results: (number[] | null)[] = [];
    for (const text of texts) {
      results.push(await this.generateEmbedding(text));
    }
    return results;
  }
}

File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/ollama/OllamaClient.kt
/** Optional Ollama HTTP client for semantic embeddings. */
package com.codeintel.ollama

import com.codeintel.log
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

@Serializable
data class EmbeddingRequest(val model: String, val prompt: String)

@Serializable
data class EmbeddingResponse(val embedding: List<Double>)

class OllamaClient(private val baseUrl: String, private val model: String) {
    private val client = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(10))
        .build()
    private val json = Json { ignoreUnknownKeys = true }

    /** Generate embedding for text. Returns null on failure. */
    fun embed(text: String): List<Double>? {
        val body = json.encodeToString(EmbeddingRequest.serializer(), EmbeddingRequest(model, text))
        val request = HttpRequest.newBuilder()
            .uri(URI.create("$baseUrl/api/embeddings"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .timeout(Duration.ofSeconds(30))
            .build()
        return try {
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() != 200) return null
            json.decodeFromString(EmbeddingResponse.serializer(), response.body()).embedding
        } catch (e: Exception) {
            log("Ollama error: ${e.message}")
            null
        }
    }

    /** Check if Ollama is reachable. */
    fun isAvailable(): Boolean {
        return try {
            val req = HttpRequest.newBuilder()
                .uri(URI.create("$baseUrl/api/tags"))
                .timeout(Duration.ofSeconds(5))
                .GET().build()
            client.send(req, HttpResponse.BodyHandlers.ofString()).statusCode() == 200
        } catch (_: Exception) { false }
    }
}
